Formula/micronaut.rb

```diff
@@ -1,17 +1,26 @@
 class Micronaut < Formula
   desc "Modern JVM-based framework for building modular microservices"
   homepage "https://micronaut.io/"
-  url "https://github.com/micronaut-projects/micronaut-core/releases/download/v1.3.6/micronaut-1.3.6.zip"
-  sha256 "271a5af14f87e78ec8c72cfd09d04c50c96826532a5496bd9651a2775815693f"
+  url "https://github.com/micronaut-projects/micronaut-starter/archive/v2.0.0.RC1.tar.gz"
+  sha256 "91dc71b200c598a91b95a20792dcc3f3cb160b8e169b581f6b98689dc713acad"
 
-  bottle :unneeded
+  GRAALVM_HOME="./graalvm/Contents/Home".freeze
 
-  depends_on "openjdk"
+  depends_on :java => ["1.8", :build]
 
   def install
-    rm_f Dir["bin/*.bat"]
-    libexec.install %W[bin media cli-#{version}.jar]
-    (bin/"mn").write_env_script libexec/"bin/mn", :JAVA_HOME => "${JAVA_HOME:-#{Formula["openjdk"].opt_prefix}}"
+    with_env "GRAAL_OS" => "darwin-amd64", "GRAAL_VERSION" => "20.1.0" do
+      system "./install-graal.sh"
```

Comment:
Does this download the graal binaries?

Reply:
Yes, it. does.

Reply:
In that case we need to use `resource` blocks. We need a versioned checksummed download.

Reply:
```
https://github.com/graalvm/graalvm-ce-builds/releases/download/vm-20.1.0/graalvm-ce-java8-darwin-amd64-20.1.0.tar.gz looks like a binary package, not a source archive; homebrew/core is source-only
```

This is becoming a nightmare loop 🤦

Reply:
Yeah, it definitely shouldn't pull binaries to build unless there is no other way. This is why I recommended making a graalvm formula.

Reply:
Making a graalvm formula is not practical, plus other JDKs aren't built from source either. Why on earth would someone want to use an unofficially-built JDK? They offer pre-compiled binaries for a reason.

Can we move forward with using their binaries?

Reply:
https://docs.brew.sh/Acceptable-Formulae#we-dont-like-install-scripts-that-download-unversioned-things

Reply:
I have tried

```ruby
  resource "graalvm" do
    url "https://github.com/graalvm/graalvm-ce-builds/releases/download/vm-20.1.0/graalvm-ce-java8-darwin-amd64-20.1.0.tar.gz"
    sha256 "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
  end
```

Which is a checksummed, versioned resource. But `audit` won't like it.

Regarding your last comment, I think we are missing some context here: the Micronaut CLI is as simple as a single `mn` binary. And I already have a formula prepared that builds such binary from source.

In this context, asking for a GraalVM formula is ridiculously out of scope. It could be never ending, because GraalVM has dozens of dependencies. Will I have to provide formulas for those missing in homebrew?

With all respect, I think the request for a GraalVM formula is unreasonable.

Reply:
I don't think it's particularly unreasonable to ask you to follow the homebrew rules for packaging. They're pretty well documented iyam. You just have the shitty luck to be the first person that wants to package something in homebrew using graalvm.

Reply:
I might be in a similar boat, I just submitted a Cask for a GraalVM based project.
A GraalVM formula does seem like it would be incredibly complicated and difficult to maintain, and using the latest version of Graal isn't always possible.  For example, something broke in the latest release that causes my project not to build (and I had to submit PRs against the project)

Homebrew/homebrew-cask#84527

Reply:
The "homebrew rules" are subject to different interpretations. In https://github.com/Homebrew/homebrew-cask/pull/84439#issuecomment-644889960 I was told to remove this formula in favour of the cask.

The formula way has hit a roadblock, no way anyone can contribute a GraalVM formula.

So please clarify: can we move forward with the cask approach?

Reply:
> no way anyone can contribute a GraalVM formula.

> A GraalVM formula does seem like it would be incredibly complicated and difficult to maintain

May I ask why do you think so?
As far as I can see there was no attempt to add it to homebrew-core (in PRs).

> So please clarify: can we move forward with the cask approach?

I believe the only way to continue — is to create a `graalvm` formula then update `micronaut`, any thoughts on this discussion @Homebrew/core? 

And you always can create a separate tap for this (for cask or for a formula https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap).

Reply:
> I believe the only way to continue — is to create a `graalvm` formula then update `micronaut`, any thoughts on this discussion @Homebrew/core?

Yes. Alternatively, if `micronaut` can work with `openjdk`: use that instead.

---

@alvarofauna 

> This is becoming a nightmare loop 🤦
> With all respect, I think the request for a GraalVM formula is unreasonable.
> The formula way has hit a roadblock, no way anyone can contribute a GraalVM formula.

We're all volunteers and doing our best here @alvarosanchez. Please be a bit more patient, tolerant and respectful towards our maintainers. We are trying our best to help you here and the rules we have exist for a reason.

Reply:
@MikeMcQuaid It wasn't my intention to be disrespectful. Please accept my apologies if it wasn't the case.

I have followed all the instructions that have been given to me, leading to 3 PRs (https://github.com/Homebrew/homebrew-cask/pull/84439, https://github.com/Homebrew/homebrew-core/pull/56440 and this one). Unfortunately, I have been given contradictory feedback. If I had known since the beginning that the only way is to provide a GraalVM formula, we all would have saved a lot of time.

Having said that, our only viable alternative is to create our own tap. Thank you all for the time spent in this discussion.

Reply:
@MikeMcQuaid I think yeah we are going to have to just settle on providing our own tap which is not ideal, a shame for Homebrew users and doesn't integrate as well with homebrew but I believe it is the most user friendly approach for our users.

The reason is, even if there was a GraalVM formula that we could use to build our CLI from source currently building a native image using GraalVM takes significant hardware resources. Sometimes users may be required to have 6-8 gigabytes of memory available to compute the native image and it can take several minutes to build even on modern hardware.

Given this it is unreasonable to ask users to build from source a native image using GraalVM on their local machines and it would represent a poor user experience IMO. This is why the cask approach seemed the best way forward, however if the Homebrew community is unwilling to tweak this policy for native binaries built with GraalVM we will have to stick with the custom tap I suppose.

Reply:
We are an open source package manager for open source software so we need to built it from source, sorry.

Reply:
I'd also like to point out that only the developer of the formula will likely ever build the formula locally. We're providing binary packages to end users that are compiled on CI so resource usage shouldn't factor into it.

Reply:
Ok I see, understood. I think this is going to be a common requirement going forward so hopefully a GraalVM native forumla that someone maintains emerges and then we can revisit the issue.

I don't think we (the Micronaut team) can be responsible for maintaining and building that.

Reply:
Anyone can make any formula they want, you could suggest to the GraalVM team to make a formula. I'm also not expecting any one person/entity to maintain any formula. Homebrew works as well as it does because the community maintains the formula.

Reply:
@MikeMcQuaid re: Alternatively, if micronaut can work with openjdk: use that instead.

OpenJDK does NOT provide an option to compile java into a single native binary (which is what is used here, and why we've chosen to use it in Homebrew/homebrew-cask#84527 as well.

> We are an open source package manager for open source software so we need to build it from source

There are exceptions to this, for example FireFox uses a cask, and many casks are commercial/private source projects

I do understand the goal of building from source when possible, I'm just not sure what qualifies for an exemption to this rule.

Sound like I need to go the Tap route as well.

Reply:
> There are exceptions to this, for example FireFox uses a cask, and many casks are commercial/private source projects

From the cask readme:

> Homebrew Cask extends Homebrew and brings its elegance, simplicity, and speed to the installation and management of GUI macOS applications such as Atom and Google Chrome.

There's also the option to apply something as a cask when it doesn't match this but can't be build from source, in general only GUI applications should go in Cask though.

Reply:
> OpenJDK does NOT provide an option to compile java into a single native binary

That doesn't matter to us, really. We can install it as a dependency. That's better because it means a single OpenJDK bump will fix a CVE in all dependants rather than requiring them all to be recompiled.

> There are exceptions to this, for example FireFox uses a cask

https://docs.brew.sh/Acceptable-Formulae#stuff-that-builds-an-app

> I do understand the goal of building from source when possible, I'm just not sure what qualifies for an exemption to this rule.

What qualifies as an exception is: the Homebrew maintainers think it should qualify for an exception (and likely change the documentation accordingly), sorry.